cpp/include/cudf/sorting.hpp

```diff
@@ -390,5 +390,78 @@ std::unique_ptr<column> top_k_order(
   rmm::cuda_stream_view stream      = cudf::get_default_stream(),
   rmm::device_async_resource_ref mr = cudf::get_current_device_resource_ref());
 
+/**
+ * @brief Computes the top k values within each segment of a column
+ *
+ * This performs the equivalent of a sort and the slice of the resulting first k elements
+ * within each segment.
+ * However, the values within each segment may not necessarily be sorted.
+ * If a segment contain less than k elements then all values for that segment are returned.
+ *
+ * @code{.pseudo}
+ * Example:
+ * col = [ 3, 4, 5, 4, 1, 2, 3, 5, 6, 7, 8, 9, 10 ]
+ * offsets = [0, 3, 7, 13]
+ * result = cudf::top_k_segmented(col, offsets, 3);
+ * result is [[5,4,3], [4,3,2], [10,8,9]] // each segment may not be sorted
+ * @endcode
+ *
+ * @throw std::invalid_argument if k less than or equal to zero
+ * @throw cudf::data_type_error if segment_offsets is not size_type
+ * @throw std::invalid_argument segments_offsets is empty or contains nulls
+ *
+ * @param col Column to compute top k
+ * @param segment_offsets Start offset index for each contiguous segment
+ * @param k Number of values to return for each segment
+ * @param sort_order DESCENDING is the largest k values (default).
+ *                   ASCENDING is the smallest k values.
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @return A column with the top k values of the input column.
+ */
+std::unique_ptr<column> top_k_segmented(
```

Comment:
It seems like our conventions would suggest this should be named `segmented_top_k`, like `segmented_sorted_order` and `segmented_sort_by_key`.